Add logging to help diagnose 'cannot fulfill operation' errors

pmorie opened the pull request:
Part of #1363; I did some investigation on this today.  I'll write up the results but I suspect that this logging would be useful to have, so wanted to PR it before it got stale.

staebler commented:
LGTM.

The logging looks fine to me. I don't know how helpful the logging will be. I did a good bit of investigation on this issue when it was first introduced with the changes for `CurrentOperation`. I think I have a good understanding of the issue already and am more than happy to discuss it with you further.

pmorie commented:
A brief write-up of what is happening on `ServiceBinding` which I suspect is representative of what's happening when we see this on other resources:

The TLDR is that if a control loop does more than a single status update if the resource it works on, that subsequent iterations of the loop might work on old versions.  Example:

- Controller starts working on ServiceBinding X at resource version N
- Controller does a status update to record the operation it's starting on X, creating resource version N+1
- API server delivers update event for resource version N+1 to controller, which queues it
- Controller does the bind operation, and updates the status of X again, creating resource version N+2
- Controller sync ends
- Controller **starts working on X resource version N+1**
- API server delivers update event for resource version N+2 to controller, which queues it
- Controller, working on N+1 does some work (resends the bind request, which should be just fine from an OSB standpoint, since the operation is supposed to be idempotent), and tries to update the status from resource version N+1, which is when we see the status update failures

Some logs that illustrate this:

<details>

```console
$ test/integration.sh -test.run=TestBasicFlowsAsync -alsologtostderr -v 6 |& grep ServiceBinding -                                                                                                                                                                              
I1025 14:00:52.639828    4824 reflector.go:202] Starting reflector *v1beta1.ServiceBinding (10s) from github.com/kubernetes-incubator/service-catalog/pkg/client/informers_generated/externalversions/factory.go:61
I1025 14:00:52.639887    4824 reflector.go:240] Listing and watching *v1beta1.ServiceBinding from github.com/kubernetes-incubator/service-catalog/pkg/client/informers_generated/externalversions/factory.go:61
I1025 14:00:57.838137    4824 controller_binding.go:84] ServiceBinding test-namespace/test-binding: received ADD/UPDATE event for: resourceVersion: 15
I1025 14:00:57.838282    4824 controller_binding.go:166] ServiceBinding "test-namespace/test-binding": beginning to process resourceVersion: 15
I1025 14:00:57.838396    4824 controller_binding.go:201] ServiceBinding "test-namespace/test-binding": Processing
I1025 14:00:57.838547    4824 controller_binding.go:284] ServiceBinding "test-namespace/test-binding": Adding/Updating
I1025 14:00:57.838494    4824 util.go:224] Waiting for binding test-namespace/test-binding condition v1beta1.ServiceBindingCondition{Type:"Ready", Status:"True", LastTransitionTime:v1.Time{Time:time.Time{sec:0, nsec:0, loc:(*time.Location)(nil)}}, Reason:"", Message:""}
I1025 14:00:57.838622    4824 controller_binding.go:950] ServiceBinding "test-namespace/test-binding": Setting condition "Ready" to False
I1025 14:00:57.838661    4824 controller_binding.go:963] ServiceBinding "test-namespace/test-binding": Setting lastTransitionTime for condition "Ready" to 2017-10-25 14:00:57.838618323 -0400 EDT
I1025 14:00:57.838700    4824 controller_binding.go:998] ServiceBinding "test-namespace/test-binding": Updating status; resourceVersion: 15
I1025 14:00:58.239785    4824 controller_binding.go:84] ServiceBinding test-namespace/test-binding: received ADD/UPDATE event for: resourceVersion: 16
I1025 14:00:58.240019    4824 controller_binding.go:1007] ServiceBinding "test-namespace/test-binding": Updated status of resourceVersion: 15; got resourceVersion: 16
I1025 14:00:58.240075    4824 controller_binding.go:844] ServiceBinding "test-namespace/test-binding": Creating/updating Secret "test-namespace/test-binding" with 2 keys
I1025 14:00:58.240162    4824 controller_binding.go:950] ServiceBinding "test-namespace/test-binding": Setting condition "Ready" to True
I1025 14:00:58.240199    4824 controller_binding.go:974] ServiceBinding "test-namespace/test-binding": Found status change for condition "Ready": "False" -> "True"; setting lastTransitionTime to 2017-10-25 14:00:58.240156901 -0400 EDT
I1025 14:00:58.240251    4824 controller_binding.go:998] ServiceBinding "test-namespace/test-binding": Updating status; resourceVersion: 16
I1025 14:00:58.438346    4824 controller_binding.go:1007] ServiceBinding "test-namespace/test-binding": Updated status of resourceVersion: 16; got resourceVersion: 17
I1025 14:00:58.438428    4824 controller_binding.go:619] ServiceBinding "test-namespace/test-binding": Successfully bound to ServiceInstance "test-namespace/test-instance" of ClusterServiceClass (K8S: "12345" ExternalName: "test-service") at ClusterServiceBroker test-brok
er
I1025 14:00:58.438646    4824 controller_binding.go:166] ServiceBinding "test-namespace/test-binding": beginning to process resourceVersion: 16
I1025 14:00:58.438694    4824 controller_binding.go:201] ServiceBinding "test-namespace/test-binding": Processing
I1025 14:00:58.438769    4824 controller_binding.go:84] ServiceBinding test-namespace/test-binding: received ADD/UPDATE event for: resourceVersion: 17
I1025 14:00:58.438787    4824 controller_binding.go:284] ServiceBinding "test-namespace/test-binding": Adding/Updating
I1025 14:00:58.438859    4824 controller_binding.go:844] ServiceBinding "test-namespace/test-binding": Creating/updating Secret "test-namespace/test-binding" with 2 keys
I1025 14:00:58.438923    4824 controller_binding.go:950] ServiceBinding "test-namespace/test-binding": Setting condition "Ready" to True
I1025 14:00:58.438956    4824 controller_binding.go:974] ServiceBinding "test-namespace/test-binding": Found status change for condition "Ready": "False" -> "True"; setting lastTransitionTime to 2017-10-25 14:00:58.438919999 -0400 EDT
I1025 14:00:58.438999    4824 controller_binding.go:998] ServiceBinding "test-namespace/test-binding": Updating status; resourceVersion: 16
I1025 14:00:58.536436    4824 util.go:224] Waiting for binding test-namespace/test-binding condition v1beta1.ServiceBindingCondition{Type:"Ready", Status:"True", LastTransitionTime:v1.Time{Time:time.Time{sec:0, nsec:0, loc:(*time.Location)(nil)}}, Reason:"", Message:""}
E1025 14:00:58.636950    4824 controller_binding.go:1004] ServiceBinding "test-namespace/test-binding": Error updating status
I1025 14:00:58.636996    4824 controller.go:232] Error syncing ServiceBinding test-namespace/test-binding: Operation cannot be fulfilled on servicebindings.servicecatalog.k8s.io "test-binding": the object has been modified; please apply your changes to the latest version 
and try again
I1025 14:00:58.637078    4824 controller_binding.go:166] ServiceBinding "test-namespace/test-binding": beginning to process resourceVersion: 17
I1025 14:00:58.637136    4824 controller_binding.go:188] ServiceBinding "test-namespace/test-binding": Not processing event; reconciled generation showed there is no work to do
I1025 14:00:58.642170    4824 controller_binding.go:166] ServiceBinding "test-namespace/test-binding": beginning to process resourceVersion: 17
I1025 14:00:58.642244    4824 controller_binding.go:188] ServiceBinding "test-namespace/test-binding": Not processing event; reconciled generation showed there is no work to do
I1025 14:00:59.040680    4824 controller_binding.go:84] ServiceBinding test-namespace/test-binding: received ADD/UPDATE event for: resourceVersion: 18
I1025 14:00:59.040763    4824 controller_binding.go:166] ServiceBinding "test-namespace/test-binding": beginning to process resourceVersion: 18
I1025 14:00:59.040815    4824 controller_binding.go:201] ServiceBinding "test-namespace/test-binding": Processing
I1025 14:00:59.040897    4824 controller_binding.go:912] ServiceBinding "test-namespace/test-binding": Deleting Secret "test-namespace/test-binding"
I1025 14:00:59.040937    4824 controller_binding.go:950] ServiceBinding "test-namespace/test-binding": Setting condition "Ready" to False
I1025 14:00:59.040975    4824 controller_binding.go:974] ServiceBinding "test-namespace/test-binding": Found status change for condition "Ready": "True" -> "False"; setting lastTransitionTime to 2017-10-25 14:00:59.040934065 -0400 EDT
I1025 14:00:59.041034    4824 controller_binding.go:998] ServiceBinding "test-namespace/test-binding": Updating status; resourceVersion: 18
I1025 14:00:59.439673    4824 controller_binding.go:1007] ServiceBinding "test-namespace/test-binding": Updated status of resourceVersion: 18; got resourceVersion: 19
I1025 14:00:59.439722    4824 controller_binding.go:950] ServiceBinding "test-namespace/test-binding": Setting condition "Ready" to False
I1025 14:00:59.439754    4824 controller_binding.go:998] ServiceBinding "test-namespace/test-binding": Updating status; resourceVersion: 19
I1025 14:00:59.439820    4824 controller_binding.go:84] ServiceBinding test-namespace/test-binding: received ADD/UPDATE event for: resourceVersion: 19
I1025 14:00:59.639331    4824 controller_binding.go:1051] ServiceBinding "test-namespace/test-binding": Received DELETE event; no further processing will occur; resourceVersion: 20
I1025 14:00:59.639443    4824 controller_binding.go:1007] ServiceBinding "test-namespace/test-binding": Updated status of resourceVersion: 19; got resourceVersion: 19
I1025 14:00:59.639516    4824 controller_binding.go:820] ServiceBinding "test-namespace/test-binding": Successfully deleted ServiceBinding of ServiceInstance "test-namespace/test-instance" of ClusterServiceClass (K8S: "12345" ExternalName: "test-service") at ClusterServic
eBroker test-broker
I1025 14:00:59.639594    4824 controller_binding.go:99] ServiceBinding "test-namespace/test-binding": Not doing work because the ServiceBinding has been deleted
```

</details>

The mitigation that we can do for this is refactor the controller sync loops so that they only do a single status update per iteration.  This means work is spread across more than 1 iteration, but each iteration is probably easier to read about.  Example of what this might mean for binding:

- Iteration 1: controller decides binding is ready to bind; marks the bind operation as in progress and returns
- Iteration 2: controller does the bind, handles secret injection, updates the binding status to ready

I'm going to play around with `ServiceBinding` and see if I can get the above flow working without a ton of changes (I think this should be possible).  I would rather not have a happy path in the controller that we know will result in duplicate operations to the broker or status update failures.

pmorie commented:
Thanks a lot to @liggitt for digging into this with me.
